DOCS-1592: Make PyInstaller recommended method and edit build command

sguequierre opened the pull request:
Implementing suggestions from [scope](https://docs.google.com/document/d/1E0EzdlBNrPXZi5ki1J5FGAiOIETYVHE0fUVtV9JN5I8/edit#heading=h.tcicyojyqi6c)

npentrel commented on docs/registry/upload/_index.md:
````diff
@@ -317,18 +334,71 @@ jobs:
 ```
 
-You can test this build configuration by running the following command on your development machine:
+{{%expand "Click to view example build-linux-arm64.sh" %}}
````
```suggestion
{{% expand "Click to view example build-linux-arm64.sh" %}}
```

npentrel commented on docs/registry/upload/_index.md:
````diff
@@ -317,18 +334,71 @@ jobs:
+tar -czvf dist/archive.tar.gz dist/main ##create archive from entrypoint
+```
+
+{{% /expand%}}
````
```suggestion
{{% /expand %}}
```

npentrel commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,71 @@ jobs:
+
+{{% /expand%}}
+
+{{%expand "Click to view example build-darwin-arm64.sh" %}}
```
```suggestion
{{% expand "Click to view example build-darwin-arm64.sh" %}}
```

npentrel commented on docs/registry/upload/_index.md:
````diff
@@ -317,18 +334,71 @@ jobs:
+tar -czvf dist/archive.tar.gz dist/main ##create archive from entrypoint
+```
+
+{{% /expand%}}
````
```suggestion
{{% /expand %}}
```

npentrel commented:
will let andrew review but added a few nits

andf-viam commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,71 @@ jobs:
 The `build-action` GitHub action relies on a build command that you need to specify in the <file>meta.json</file> file that you created for your module when you first [uploaded it](/registry/upload/#upload-a-custom-module-using-the-cli).
 At the end of your <file>meta.json</file>, add the build configuration:
 
+{{< tabs >}}
```
Tabbed setup looks great!

andf-viam commented on docs/fleet/cli.md:
````diff
@@ -702,18 +702,69 @@ You can use the `module build start` or `module build local` commands to build y
+```json {class="line-numbers linkable-line-numbers"}
+"build": {
+  "path" : "dist/archive.tar.gz",               // optional - path to your built module
+                                          // (passed to the 'viam module upload' command)
````
Supernit: comment alignment:

```suggestion
                                                // (passed to the 'viam module upload' command)
```

andf-viam commented on docs/fleet/cli.md:
```diff
@@ -702,18 +702,69 @@ You can use the `module build start` or `module build local` commands to build y
   "setup": "./setup.sh",                  // optional - command to install your build dependencies
-  "build": "make module.tar.gz",          // command that will build your module
-  "path" : "module.tar.gz",               // optional - path to your built module
+  "build": "sh build.sh",          // command that will build your module
```
Supernit: comment alignment:

```suggestion
  "build": "sh build.sh",                 // command that will build your module
```

andf-viam commented on docs/fleet/cli.md:
```diff
@@ -702,18 +702,69 @@ You can use the `module build start` or `module build local` commands to build y
-  "build": "make module.tar.gz",          // command that will build your module
-  "path" : "module.tar.gz",               // optional - path to your built module
+  "build": "sh build.sh",          // command that will build your module
+  "path" : "dist/archive.tar.gz",               // optional - path to your built module
```
```suggestion
  "path" : "dist/archive.tar.gz",         // optional - path to your built module
```

andf-viam commented on docs/fleet/cli.md:
```diff
@@ -702,18 +702,69 @@ You can use the `module build start` or `module build local` commands to build y
+{{% /tab %}}
+{{< /tabs >}}
+
 For example, the following extends the `my-module` <file>meta.json</file> file from the previous section with a new `build` object to control its build parameters when used with `module build start` or `module build local`:
```
Nit: Now that we have two build approaches, let's clarify which one this is an example of:

```suggestion
For example, the following extends the `my-module` <file>meta.json</file> file from the previous section using the single build file approach, adding a new `build` object to control its build parameters when used with `module build start` or `module build local`:
```

andf-viam reviewed (approved):
LGTM with some small nits. Great work!

andf-viam reviewed (commented):
Oh, and just to answer your question in the PR desc: I believe you've fully captured the ask! Deleting all non-`PyInstaller` build methods seems to be the right call to me to encourage 85% adoption for `PyInstaller`. Agreed with eng that is simplifies the build steps, esp for multi-platform modules, and is a solid choice to standardize on as _the way_. 👍

sguequierre commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,36 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
 
-Compiling your Python module in this fashion ensures that your module has access to any packages it requires during runtime.
-If you intend to share your module with other users, or to deploy it to a fleet of machines, this approach "bundles" your module code together with its required dependencies, making your module highly-portable across like architectures.
+   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the base name of the input script (in this case, main).
```
```suggestion
   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).
```

sguequierre commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,36 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
+Follow the instructions to [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action) to add the build configuration to your machine.
+
+With this approach, you can make a build script like the following to
+build your module, and configure the resulting executable (<file>dist/main</file>) as your module `"entrypoint"`:
```
```suggestion
build your module:
```

sguequierre commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,36 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
 
-Compiling your Python module in this fashion ensures that your module has access to any packages it requires during runtime.
-If you intend to share your module with other users, or to deploy it to a fleet of machines, this approach "bundles" your module code together with its required dependencies, making your module highly-portable across like architectures.
+   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).
```
@andf-viam added all of this!

sguequierre commented on docs/registry/upload/_index.md:
```diff
@@ -289,8 +306,8 @@ However, if you already have your own CI with access to arm runners or only inte
 
-   {{< tabs >}}
-   {{% tab name="CI with build-action" %}}
+{{< tabs >}}
```
@andf-viam do you have any idea why it keeps telling me this tab is unclosed and failing to build on hugo when AFAIK it is closed? I keep checking and I can't find where I changed the tabs to be wrong but they are messed up

sguequierre commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,103 @@ jobs:
+
+{{% /expand%}}
+
+{{%expand "Click to view example build-darwin-arm64.sh" %}}
```
that fixed it

sguequierre commented on docs/registry/upload/_index.md:
````diff
@@ -158,19 +175,19 @@ For more information, see [Naming your model](/registry/#naming-your-model-names
    ```
 
-   Where `run.sh` is your [entrypoint file](/registry/create/#compile-or-package-your-module), `requirements.txt` is your [pip dependency list file](/registry/create/#compile-or-package-your-module), and `src` is the source directory of your module.
+  In this case, `dist/main` is the [packaged executable](/registry/create/#compile-or-package-your-module) that runs the module at the [entry point](/registry/create/#write-an-entry-point-main-program-file).
````
```suggestion
 Where `dist/main` is the [packaged executable](/registry/create/#compile-or-package-your-module) that runs the module at the [entry point](/registry/create/#write-an-entry-point-main-program-file).
```

sguequierre commented on docs/fleet/cli.md:
````diff
@@ -702,19 +702,103 @@ You can use the `module build start` or `module build local` commands to build y
 
 ```json {class="line-numbers linkable-line-numbers"}
 "build": {
   "setup": "./setup.sh",                  // optional - command to install your build dependencies
````
@michaellee1019 thinking including the `setup` line doesn't make sense here if build dependencies are installed with the build script but still want to show the option-- do you think that's ok? or would it be possible to have both?

michaellee1019 replied:
Yes both are supported at once. You could separate the file below to move the code up to `pip3 install -r requirements.txt` and then start the build script with `python3 -m PyInstaller...`. Up to you. It is better to have separate files for separation of concerns but likely easier for developers to copy/paste one file.

sguequierre replied:
Ok thanks so much @michaellee1019 -- going to show both methods since the expand hides unless they select so doesn't take up too much room

michaellee1019 commented:
Just checking: Are we sure we want to do this now? Last time we discussed in a review meeting, we wanted to try out on a few viam supported modules and monitor for a while before doing this. I might be out of date with recent decisions here. So far I know we have implemented it on one that found some build related issues. @abe-winter I believe has fixed the issues.

michaellee1019 commented:
> So far I know we have implemented it on one that found some build related issues. @abe-winter I believe has fixed the issues.

Sounds like we need more testing to be done to ensure the build works correctly. I think we just need to test on a few modules and then are good.

sguequierre commented on docs/registry/create/_index.md:
````diff
@@ -1381,13 +1294,36 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
+tar -czvf dist/archive.tar.gz dist/main
+```
+
+This script automates the process of setting up a Python virtual environment, installing dependencies, packaging the Python module into a standalone executable using PyInstaller, and then compressing the resulting executable into a tarball.
````
```suggestion
This script automates the process of setting up a Python virtual environment on a Linux arm64 machine, installing dependencies, packaging the Python module into a standalone executable using PyInstaller, and then compressing the resulting executable into a tarball.
For more examples of build scripts see [Update an existing module using a GitHub action](/registry/upload/#update-an-existing-module-using-a-github-action).
```

sguequierre commented:
> > So far I know we have implemented it on one that found some build related issues. @abe-winter I believe has fixed the issues.
> 
> Sounds like we need more testing to be done to ensure the build works correctly. I think we just need to test on a few modules and then are good.

Okay! Would you be able to let us know and review this PR once that happens? I didn't realize we weren't sure about doing this yet, just picked up this ticket from our TODO in Jira. Happy to wait to merge as long as you think necessary. I did talk to Abe about his fix for the python versioning thing/suggestions for the build scripts and incorporated them.

abe-winter commented:
sorry, this has been blocked on me -- the fix for this is in review,
hopefully out this week 🤞🙏. will update here when done

sguequierre commented:
> sorry, this has been blocked on me -- the fix for this is in review, hopefully out this week 🤞🙏. will update here when done

no worries!! thank you!!

abe-winter commented:
cloud-build python issues fixed! clear to proceed, lmk if I can help

michaellee1019 commented on docs/fleet/cli.md:
````diff
@@ -702,19 +702,103 @@ You can use the `module build start` or `module build local` commands to build y
+{{% tab name="Platform Specific" %}}
+
+```json {class="line-numbers linkable-line-numbers"}
+"build": {
+  "path" : "dist/archive.tar.gz",               // optional - path to your built module
+                                                // (passed to the 'viam module upload' command)
+  "arch": {
+        "linux/arm64": {
+          "build": "sh build-linux-arm64.sh" // command that will build your module
+        },
+        "darwin/arm64": {
+          "build": "sh build-darwin-arm64.sh" // command that will build your module
+        }
+      } // architecture(s) to build for
+}
````
This isn't supported yet. Will have to do that if check above for now. This platform level build script is in our plan but not built yet.

sguequierre replied:
Ah gotcha, apologies for the confusion-- this ticket was linked to scope

michaellee1019 commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,37 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
+{{% alert title="Note" color="note" %}}
+
+Used in this manner, PyInstaller does not support relative imports (imports starting with `.`).
 In addition, PyInstaller does not support cross-compiling: you must compile your module on the target architecture you wish to support.
 For example, you cannot run a module on a Linux `arm64` system if you compiled it using PyInstaller on a Linux `amd64` system.
```
Wondering if we should rephrase and emphasize using our built system.

"Viam makes this easy for you by providing a build system for modules. Simply follow these instructions in our CLI docs to automatically build for each system your module can support."

michaellee1019 commented on docs/registry/upload/_index.md:
```diff
@@ -143,7 +149,18 @@ If you mark your module as public, you cannot change it back to private.
-      "entrypoint": "my-module.sh"
+      "build": {
+        "path": "dist/archive.tar.gz",
+        "arch": {
+          "linux/arm64": {
+            "build": "sh build-linux-arm64.sh" // command that will build your module
+          },
+          "darwin/arm64": {
+            "build": "sh build-darwin-arm64.sh" // command that will build your module
+          }
+        } // architecture(s) to build for
```
Again, not supported yet.

michaellee1019 commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,105 @@ jobs:
+  "module_id": "example-module",
+  ...
+  "build": {
+    "arch": {
+          "linux/arm64": {
+            "path" : "dist/archive.tar.gz",               // optional - path to your built module
+            "build": "sh build-linux-arm64.sh" // command that will build your module
+          },
+          "darwin/arm64": {
+            "build": "sh build-darwin-arm64.sh" // command that will build your module
+          }
+        } // architecture(s) to build for
+  }
```
One more.

abe-winter commented on docs/fleet/cli.md:
```diff
@@ -702,19 +702,103 @@ You can use the `module build start` or `module build local` commands to build y
   "setup": "./setup.sh",                  // optional - command to install your build dependencies
-  "build": "make module.tar.gz",          // command that will build your module
-  "path" : "module.tar.gz",               // optional - path to your built module
+  "build": "sh build.sh",                 // command that will build your module
```
would change this to `./build.sh` unless someone objects. doing it this way runs it under the `sh` shell, which is weird because the shebang in 'example build.sh' below uses bash

(sh doesn't support the `source` command, which we use sometimes)

sguequierre replied:
```suggestion
  "build": "./build.sh",                 // command that will build your module
```

abe-winter commented on docs/fleet/cli.md:
````diff
@@ -702,19 +702,103 @@ You can use the `module build start` or `module build local` commands to build y
+}
+```
+
+{{%expand "Click to view example build-linux-arm64.sh" %}}
````
non critical, but this `build-linux` + the `build-darwin` below could be merged -- can do the `if uname` thing from the previous script

sguequierre replied:
Commenting out this whole section re. Michael's comment, but I meant to show the separate files for this approach!

abe-winter commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,37 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
-If you intend to share your module with other users, or to deploy it to a fleet of machines, this approach "bundles" your module code together with its required dependencies, making your module highly-portable across like architectures.
+   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).
+
+We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86, ARM Linux, and MacOS.
```
platforms list: 'x86 and Arm Linux, and MacOS' (or something)

abe-winter commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,37 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
+
+{{% alert title="Note" color="note" %}}
+
+Used in this manner, PyInstaller does not support relative imports (imports starting with `.`).
```
inability to do relative imports could alienate people who are heavy python users

consider changing to something like:

```suggestion
Note that PyInstaller does not support relative imports in entrypoints (imports starting with `.`). If you get 'ImportError: attempted relative import with no known parent package', set up a stub entrypoint as described [here](https://github.com/pyinstaller/pyinstaller/issues/2560).
```

sguequierre replied:
awesome thanks!

abe-winter commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,105 @@ jobs:
   "build": {
-    "build": "make module.tar.gz",
-    "arch" : ["linux/amd64", "linux/arm64", "darwin/arm64"]
+    "build": "sh build.sh",          // command that will build your module
```
as above, would do `./build.sh` to fix sh / bash issue

sguequierre replied:
Thank you! yeah sorry I didn't implement earlier, was slightly confused about if I had to instruct them to make it executable or do something else differently this way

abe-winter commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,105 @@ jobs:
+    "arch": {
+          "linux/arm64": {
+            "path" : "dist/archive.tar.gz",               // optional - path to your built module
+            "build": "sh build-linux-arm64.sh" // command that will build your module
```
`./build-x.sh` instead of `sh build` here + below

sguequierre replied:
```suggestion
            "build": "./build-linux-arm64.sh" // command that will build your module
```

abe-winter commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,105 @@ jobs:
+#!/bin/bash
+set -e
+
+sudo apt-get install -y python3-venv
```
could use bash `if` statement to unify this one and below -- only difference is apt-get vs brew

sguequierre replied:
Not addressing as commenting out

abe-winter reviewed (approved):
makes sense to me -- left notes

sguequierre commented on docs/registry/upload/_index.md:
```diff
@@ -317,18 +334,105 @@ jobs:
+            "build": "sh build-linux-arm64.sh" // command that will build your module
+          },
+          "darwin/arm64": {
+            "build": "sh build-darwin-arm64.sh" // command that will build your module
```
```suggestion
            "build": "./build-darwin-arm64.sh" // command that will build your module
```

sguequierre commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,38 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
+{{% alert title="Note" color="note" %}}
+
+Note that PyInstaller does not support relative imports in entrypoints (imports starting with `.`).
+If you get 'ImportError: attempted relative import with no known parent package', set up a stub entrypoint as described [here](https://github.com/pyinstaller/pyinstaller/issues/2560).
```
```suggestion
If you get 'ImportError: attempted relative import with no known parent package', set up a stub entrypoint as described on [GitHub](https://github.com/pyinstaller/pyinstaller/issues/2560).
```

sguequierre commented on docs/registry/create/_index.md:
```diff
@@ -1381,13 +1294,38 @@ Use the [`PyInstaller` package](https://pypi.org/project/pyinstaller/) to compil
-If you intend to share your module with other users, or to deploy it to a fleet of machines, this approach "bundles" your module code together with its required dependencies, making your module highly-portable across like architectures.
+   By default, the output directory for the packaged executable is <file>dist</file>, and the name of the executable is derived from the name of the input script (in this case, main).
+
+We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86, ARM Linux, and MacOS.
```
```suggestion
We recommend you use PyInstaller with the [`build-action` GitHub action](https://github.com/viamrobotics/build-action) which provides a simple cross-platform build setup for multiple platforms: x86 and Arm Linux, and MacOS.
```

sguequierre commented:
@michaellee1019 Do you think this is ready to merge?

michaellee1019 reviewed (approved):
Yes lets merge this in. Excited to see this improvement happening.
